### ISSUE TITLE
Filter on a field leads to optimization error

### ISSUE THREAD

**Issue by Mapiarz** [blb-ventures/strawberry-django-plus#192]
Hi! I'm looking to upgrade from strawberry-django-plus `1.35.2` to `2.1.0`. I'm seeing some regressions.

For example, given this schema:
```python
@gql.django.type(inventory.InventoryItem)
class InventoryItem(relay.Node):
    name: gql.auto
    in_manual: gql.auto
    images: list["InventoryItemImage"]


@gql.django.type(inventory.InventoryItemImage)
class InventoryItemImage(gql.Node):
    @gql.django.field
    def file(self) -> str:
        return self.file.url


@gql.django.type(inventory.InventoryListGroup2InventoryItem)
class InventoryItemInGroup(gql.Node):
    item: InventoryItem
    quantity: gql.auto


@gql.django.filter(inventory.InventoryListGroup2InventoryItem, lookups=True)
class InventoryItemInGroupFilter:
    in_manual: Optional[bool]

    def filter_in_manual(self, queryset: QuerySet[inventory.InventoryListGroup2InventoryItem]):
        if not self.in_manual:
            return queryset
        return queryset.filter(item__in_manual=self.in_manual)


@gql.django.type(inventory.InventoryListGroup, filters=InventoryListGroupFilter)
class InventoryListGroup(gql.Node):
    name: gql.auto
    items_through: List[InventoryItemInGroup] = gql.django.field(filters=InventoryItemInGroupFilter, name="items")
```

And this query:
```graphql
{
  node(
    id: "SW52ZW50b3J5TGlzdEdyb3VwOjgxYzMwM2Q0LTY5ODYtNDU3OS05MGQ4LWNjMmQ3M2NkNzMzMQ=="
  ) {
    ... on InventoryListGroup {
      items(filters: {inManual: true}) {
        item {
          images {
            file
          }
        }
      }
    }
  }
}
```

On `1.35.2` this query works as intended. On `2.1.0` it fails with `Optimization error: lookup was already seen with a different queryset. You may need to adjust the ordering of your lookups`. If I don't filter the items in the query, it works in `2.1.0` as well.

What's interesting, when investigating with DBT, I can see with the filter in the query, there's indeed a duplicated query. What's weird about it is that the duplicated query is there in both `1.35.2` and `2.1.0` but only the new version fails to return a result with the aforementioned error.

This seems possibly related to https://github.com/blb-ventures/strawberry-django-plus/issues/159?

Please advise on how to proceed. Unsure if I'm dealing with an outright bug here or some limitation. Eitherway, please let me know if there are some workarounds I could apply.

**Comment by bellini666** [comment]
Hey @Mapiarz ,

The main issue here, just like in #159 , is that the optimizer is doing a prefetch for all your `items_through` and them you throw them away when you do a filter on top of item.

Take a look at [this comment](https://github.com/blb-ventures/strawberry-django-plus/issues/159#issuecomment-1368013678) for more information and how to workaround that.

I wonder if the `Optimization error: lookup was already seen with a different queryset. You may need to adjust the ordering of your lookups` is a side effect of that or really an unrelated issue? There are a lot of differences between 1.35.2 to 2.1.0, but the only breaking one is this: https://github.com/blb-ventures/strawberry-django-plus/releases/tag/v2.0

**Comment by Mapiarz** [comment]
Thanks @bellini666. 

For the workaround, do you have any idea how to get the filter value in a prefetch lambda with info? Unless I’m missing something I have to manually parse all the nodes and figure out the filter argument node. 

I would also like to know what the baseline here is. Is this something that you plan on looking into or are you saying this is unsupported and we have to live with it? If not supported then at least we should try to establish and document what is the scope of this limitation. Basically I’m trying to figure out the best path forward for me, which could be:
- wait for fix
- Implement workarounds
- Restructure api not to have nested paginations/filters?

**Comment by jihwanBV** [comment]
**I used a translator.**
There is no problem until version 2.0.6, but the same error occurs as soon as version 2.1.0 is used.

**Comment by bellini666** [comment]
> For the workaround, do you have any idea how to get the filter value in a prefetch lambda with info? Unless I’m missing something I have to manually parse all the nodes and figure out the filter argument node.

There isn't an easy way right now, unfortunately. But what we could do here is change the callable you can pass to `prefetch_related` to instead or receiving just `info`, it also sends any extra arguments. So you can use `**kwargs` and retrieve the filter from `kwargs["filters"]`. This is a breaking change though, so I'll try to add something to introspect the function to check if it can accept more extra arguments.

> I would also like to know what the baseline here is. Is this something that you plan on looking into or are you saying this is unsupported and we have to live with it? If not supported then at least we should try to establish and document what is the scope of this limitation. Basically I’m trying to figure out the best path forward for me, which could be:

I definitely will look at it. Probably this weekend if I have the time.

My previous suggestions were mostly regarding the fact that your current code has n+1 problems in it because your are doing `.filter` in a related manager, which throws anything that was already prefetched out.

Something that would solve the issue for you would be to introduce an `skip_optimization` to the field, meaning it would not try to optimize it. In your current code it would be actually an optimization, because you are already not using it, and the optimization is just doing an extra prefetch in that case.

The proper fix would be to have the optimizer consider the filters and ordering passed to the field. Don't know how hard would be to implement that, but if it is not too hard I'll try that solution!

> There is no problem until version 2.0.6, but the same error occurs as soon as version 2.1.0 is used.

2.1.0 fixed some issues where some fragment spreading was not being optimized, and also connections were not being optimized since 2.0.1 I believe

So basically you were not seeing the issue because the optimizer was not actually working before =P. But I'll try to fix it properly now

**Comment by Mapiarz** [comment]
@bellini666 thanks for clearing the situation up and your ongoing support for this awesome package!

**Comment by bellini666** [comment]
@Mapiarz I'm having a hard time trying to reproduce the issue. Could you provide a minimal reproduction example so that I can run it here and try to fix?

**Comment by bellini666** [comment]
@yergom @Mapiarz I think I found the issue and it should be fixed in this release: https://github.com/blb-ventures/strawberry-django-plus/releases/tag/v2.2.0

Please tell me if that's not the case so I can investigate it further!

**Comment by yergom** [comment]
@bellini666 It works! Thanks a lot! Just to be clear for future readers, the optimizer is still not optimizing nested filters/orderings, but at least it does not throw a runtime exception any more.

**Comment by vitosamson** [comment]
Hey @bellini666 , sorry to resurrect an older issue but do you have any further updates on this idea you mentioned earlier?

> what we could do here is change the callable you can pass to prefetch_related to instead or receiving just info, it also sends any extra arguments

I ran into this today and also found it odd that the resolver itself receives an `info` that's scoped to that specific field (i.e. I can directly access `info.arguments` for that field), but the `info` that's passed to the `prefetch_related` method appears to be the top-level info which would require me to iterate through all the field nodes etc to get to the argument.

I looked through the release notes but didn't notice it in any of the newer versions.

**Comment by bellini666** [comment]
Hey @vitosamson ,

> sorry to resurrect an older issue

Np at all! :)

> I looked through the release notes but didn't notice it in any of the newer versions.

This is actually being like that since that feature was added. It is done in this line: https://github.com/blb-ventures/strawberry-django-plus/blob/main/strawberry_django_plus/optimizer.py#LL501C45-L501C45

I understand what you mean there. What do you think of maybe passing a custom object there which not only has a `.info` (which is the current info being passed), but also some extra introspection information to help in the prefetching there?

**Comment by vitosamson** [comment]
Sure, that sounds like a better solution than continually adding more args to the prefetch callable as more features are added. Something like this maybe?

```py
class FieldOptimizationContext:
    info: Info  # has access to context etc
    field: SelectedField  # has access to args etc for this specific field
```

At the least it would be great to have access to the `SelectedField` object for the field we're optimizing.

**Comment by bellini666** [comment]
@vitosamson I'm currently in the process of merging this repo into https://github.com/strawberry-graphql/strawberry-graphql-django

Most of the code is already merged there, including the optimizer, and some are waiting for review. I'll probably make a new release there this week and mark this lib as deprecated (because everything here will be available directly there)

Could you open an issue there with that request so that we can discuss the implementation there?
